### PR TITLE
Fix bug where manifests are not cached. Fixes #2770

### LINKED ISSUE #2770: Always cache miss for manifests
Checklist:

* [x] I've searched in the docs and FAQ for my answer: http://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

It looks like we always get cache misses for manifests.

**To Reproduce**

Look at logs.

**Version**

```shell
v1.3.0
```

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Optional. My organization is added to the README.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

Fixes #2770 and maybe #2351

### Changes

1. When running repo ops, if the revision is resolved, then we try return the cached version without resolving the revision.
1. Lock on chart for Helm.
1. Replaces nil `ManifestResponse` for cache with non-nil value.
1. Made cache miss/hit/error logging specific to what is being cached.
1. Meta-data request mandate resolved revision. UI only impact.

### Outcomes

1. Faster reconciliation times.
1. `test` job 1m30 faster
1. `e2e` job 2m faster

### CODE REVIEW

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -162,27 +163,24 @@ func (s *Service) runRepoOperation(
 		}
 		defer util.Close(closer)
 		return operation(chartPath, revision)
-	}
-	s.repoLock.Lock(gitClient.Root())
-	defer s.repoLock.Unlock(gitClient.Root())
-
-	if !settings.noCache && getCached(revision) {
```
Comment (alexec) [inline reposerver/repository/repository.go]: check to see if we need this

File: reposerver/repository/repository.go [hunk 2]
```diff
@@ -138,14 +138,15 @@ func (s *Service) runRepoOperation(
...
-		err = settings.sem.Acquire(c, 1)
+		err = settings.sem.Acquire(ctx, 1)
 		if err != nil {
 			return err
 		}
 		defer settings.sem.Release(1)
 	}
 
 	if source.IsHelm() {
+		// TODO - need locking here?
```
Comment (alexec) [inline reposerver/repository/repository.go]: fix here or elsewhere?

File: reposerver/repository/repository.go [hunk 3]
```diff
@@ -603,17 +601,17 @@ func runConfigManagementPlugin(appPath string, envVars *v1alpha1.Env, q *apiclie
 }
 
 func (s *Service) GetAppDetails(ctx context.Context, q *apiclient.RepoServerAppDetailsQuery) (*apiclient.RepoAppDetailsResponse, error) {
-	var res apiclient.RepoAppDetailsResponse
+	res := &apiclient.RepoAppDetailsResponse{}
 	getCached := func(revision string) bool {
 		err := s.cache.GetAppDetails(revision, q.Source, &res)
 		if err == nil {
-			log.Infof("manifest cache hit: %s/%s", revision, q.Source.Path)
```
Comment (alexec) [inline reposerver/repository/repository.go]: made logging clearer

File: reposerver/repository/repository.go [hunk 4]
```diff
@@ -130,22 +144,24 @@ func (s *Service) runRepoOperation(
...
-		err = settings.sem.Acquire(c, 1)
+		err = settings.sem.Acquire(ctx, 1)
 		if err != nil {
 			return err
 		}
 		defer settings.sem.Release(1)
 	}
 
 	if source.IsHelm() {
+		s.repoLock.Lock(source.Chart)
```
Comment (alexec) [inline reposerver/repository/repository.go]: missing lock

File: reposerver/repository/repository.go [hunk 5]
```diff
@@ -106,18 +106,32 @@ type operationSettings struct {
...
 	repo *v1alpha1.Repository,
 	source *v1alpha1.ApplicationSource,
 	getCached func(revision string) bool,
 	operation func(appPath string, revision string) error,
 	settings operationSettings) error {
 
+	revision = util.FirstNonEmpty(revision, source.TargetRevision)
+
+	// attempt to get from cache if resolved revision, as resolving is expensive
+	isResolvedRevision := false
```
Comment (alexec) [inline reposerver/repository/repository.go]: Fix (1)

File: reposerver/repository/repository.go [hunk 6]
```diff
@@ -162,27 +178,24 @@ func (s *Service) runRepoOperation(
...
+		}
+		return operation(appPath, revision)
 	}
-	return operation(appPath, revision)
 }
 
-func (s *Service) GenerateManifest(c context.Context, q *apiclient.ManifestRequest) (*apiclient.ManifestResponse, error) {
-	var res *apiclient.ManifestResponse
+func (s *Service) GenerateManifest(ctx context.Context, q *apiclient.ManifestRequest) (*apiclient.ManifestResponse, error) {
+	res := &apiclient.ManifestResponse{}
```
Comment (alexec) [inline reposerver/repository/repository.go]: Fix (3)

File: reposerver/repository/repository.go [hunk 7]
```diff
@@ -708,32 +721,33 @@ func (s *Service) GetAppDetails(ctx context.Context, q *apiclient.RepoServerAppD
...
 
-	return &res, err
+	return res, err
 }
 
 func (s *Service) GetRevisionMetadata(ctx context.Context, q *apiclient.RepoServerRevisionMetadataRequest) (*v1alpha1.RevisionMetadata, error) {
-	gitClient, commitSHA, err := s.newClientResolveRevision(q.Repo, q.Revision)
-	if err != nil {
-		return nil, err
+	if !git.IsCommitSHA(q.Revision) {
```
Comment (alexec) [inline reposerver/repository/repository.go]: Fix (5)

File: reposerver/repository/repository.go (hunk 4 above)
Comment (alexmt) [reply reposerver/repository/repository.go]: Lock is required only for git. Helm chart is exported into temp directory every time and does not require locking

File: reposerver/repository/repository.go (hunk 1 above)
Comment (alexmt) [reply reposerver/repository/repository.go]: Yes, this is still needed. The cache could be already populated after the repo lock is acquired, so we need to check it second time. https://en.wikipedia.org/wiki/Double-checked_locking

File: reposerver/repository/repository.go (hunk 6 above)
Comment (alexmt) [reply reposerver/repository/repository.go]: Can you explain please what issue is this change fixing?

File: reposerver/repository/repository.go [hunk 8]
```diff
@@ -803,6 +807,9 @@ func (s *Service) newClientResolveRevision(repo *v1alpha1.Repository, revision s
 
 func (s *Service) newHelmClientResolveRevision(repo *v1alpha1.Repository, revision string, chart string) (helm.Client, string, error) {
```
Comment (alexmt) [inline reposerver/repository/repository.go]: Please move this code into helm client and reuse in https://github.com/argoproj/argo-cd/blob/a6da0ca65b71472efedebe22afc338aeed9e268d/server/application/application.go#L1036

File: test/e2e/fixture/app/consequences.go [hunk 9]
```diff
@@ -78,12 +76,3 @@ func (c *Consequences) resource(kind, name string) ResourceStatus {
 		},
 	}
 }
-
```
Comment (alexmt) [inline test/e2e/fixture/app/consequences.go]: Is this change intentional? We don't need `ARGOCD_E2E_EXPECT_TIMEOUT` any more?

**Review by alexmt — CHANGES_REQUESTED** [review]

File: reposerver/repository/repository.go (hunk 4 above)
Comment (alexec) [reply reposerver/repository/repository.go]: you're quite right, i noted this after and did not revert

File: test/e2e/fixture/app/consequences.go (hunk 9 above)
Comment (alexec) [reply test/e2e/fixture/app/consequences.go]: not needed

File: reposerver/repository/repository.go (hunk 8 above)
Comment (alexec) [reply reposerver/repository/repository.go]: out of scope of this PR

File: reposerver/repository/repository.go [hunk 10]
```diff
@@ -794,6 +795,9 @@ func (s *Service) newClientResolveRevision(repo *v1alpha1.Repository, revision s
 	if err != nil {
 		return nil, "", err
 	}
+	if git.IsCommitSHA(revision) {
```
Comment (alexmt) [inline reposerver/repository/repository.go]: The same check is in LsRemote method: https://github.com/argoproj/argo-cd/blob/bbfb96cb0161b1d11cbbc334fb33eb1158b2c075/util/git/client.go#L334 . So this check is not required

File: server/application/application.go [hunk 11]
```diff
@@ -1033,6 +1033,9 @@ func (s *Server) resolveRevision(ctx context.Context, app *appv1.Application, sy
 		if err != nil {
 			return "", "", err
 		}
+		if helm.IsVersion(revision) {
```
Comment (alexmt) [inline server/application/application.go]: Did you mean `if helm.IsVersion(ambiguousRevision) {` ?

**Review by alexmt — CHANGES_REQUESTED** [review]

File: reposerver/repository/repository.go (hunk 8 above)
Comment (alexec) [reply reposerver/repository/repository.go]: this is tech debt, so I think out of scope

**Review by alexmt — APPROVED** [review]

**Pull request merged**